netscaler/netscaler_resource.go

```diff
@@ -124,15 +163,27 @@ func (c *NitroClient) doHTTPRequest(method string, url string, bytes *bytes.Buff
 		return []byte{}, err
 	}
 	log.Println("[DEBUG] go-nitro: response Status:", resp.Status)
-	return respHandler(resp)
+	body, err := respHandler(resp)
+	// Clear sessionid in case of session-expiry
+	if resp.Status == "401 Unauthorized" {
+		var data map[string]interface{}
+		err2 := json.Unmarshal(body, &data)
+		if err2 == nil {
+			data["errorcode"] = int(data["errorcode"].(float64))
+			if data["errorcode"] == nsErrSessionExpired || data["errorcode"] == nsErrAuthTimeout {
+				c.sessionid = ""
```

Comment:
Seems like this state management is making this client instance not thread-safe. Let's say 2 threads have reference to the same client: one reads session-id as valid. But by the time it makes the call, the second client has gotten a session timeout error and re-logs-in and swaps in a new session id. But the first call goes through and the first thread gets a session expired timeout and re-logs-in again...

Reply:
Abstracted the update/clear/get of sessionid into functions and made them thread-safe by using sync.RWMutex.